### ISSUE TITLE
Shadow's epoll and poll don't support `EPOLLERR` and `EPOLLHUP`

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#2181]
Events for poll and epoll include `EPOLLHUP`/`POLLHUP` and `EPOLLERR`/`POLLERR` to indicate hang up / error conditions on files. Shadow currently only supports `EPOLLIN` and `EPOLLOUT` for level-triggered events, so it does not correctly indicate these hang up / error conditions.

We also need better tests for these conditions, and for the behaviour of `select()`. I think sometimes a file where epoll would return `EPOLLHUP` and *not* `EPOLLIN` will still appear in the `readfds` list for `select()`.

**Comment by sporksmith** [comment]
#3574 adds a related non-shadow-passing test to `test_epoll.rs`.

For pipes, `EPOLLHUP` should be set *instead of* `EPOLLIN`

**Comment by sporksmith** [comment]
#3574 adds support for the related condition `EPOLLRDHUP`. However, we should be returning `EPOLLHUP` instead if the socket has already been shut down. 

https://github.com/shadow/shadow/pull/3574/files#r2062725974

**Comment by stevenengler** [comment]
> For pipes, EPOLLHUP should be set instead of EPOLLIN

/cc @robgjansen since you might also have thoughts about this. The problem is that the epoll code assumes that if a file becomes readable (and gets the `FileState::READABLE` state), then a blocked epoll waiting on `EPOLLIN` should be unblocked. But in this case for pipes, if the write end closes and the read end becomes readable (due to EOF) it seems we *don't* want to return `EPOLLIN` from a waiting epoll. So I think this means we need to rethink a bit how the epoll watch states work.

**Comment by robgjansen** [comment]
> So I think this means we need to rethink a bit how the epoll watch states work.

Not really?

If the `FileState` has the correct state listed:

```

**Comment by stevenengler** [comment]
> I.e., if the pipe has reached the state where "the write end closes and the read end becomes readable", then the pipe should set its FileState to HUP instead of RDHUP, or... maybe I mean it should stop setting `FileState::READABLE` in this case.

Currently a blocked `read()` uses a syscall condition that waits on `FileState::READABLE` to wake up. If we stop setting `READABLE` for pipes when the write end closes, and only set `HUP`, then `read()` will need to wait on both `READABLE` and `HUP`. Is this correct for all file types? Should `read()` wait on `READABLE`, `HUP`, and `RDHUP`?

**Comment by stevenengler** [comment]
> Is there a way to test for this in Linux? If so, would we want to test every syscall that might be sensitive to HUP/RDHUP while blocked so that we match the Linux behavior?

Yeah, I think we'd need to test all the related syscalls like `read()` and `recv()` (assuming the others like `readv()`, `recvmsg()`, etc will all be handled the same), and with different file types like pipes, tcp sockets, regular files, etc to be sure. But it's probably not practical to test all of these, and I'm not sure all of the cases where these files should be `HUP` or `RDHUP`. For example I think TCP sockets are HUP before you `connect()` or before you `listen()`, and also become HUP again once it closes for both reading and writing. And if the read end closes first then it becomes `RDHUP`.

> Well I think it is easier to adjust the `SyscallCondition` code because it's our own internal facility so we have more flexibility compared to `Epoll` where we are trying to conform to the expected libc behavior.

Okay I think I agree. We can make shadow's `FileState` just represent the Linux epoll results, and then our syscall handlers can try to construct a `SyscallCondition` that works for all file types (so I think one that waits for a change to any of READABLE, HUP, and RDHUP).
